org/mozilla/jss/provider/javax/crypto/JSSKeyManager.java

```diff
@@ -69,6 +69,16 @@ public String chooseServerAlias(String keyType, Principal[] issuers, Socket sock
         return null;  // not implemented
     }
 
+    public org.mozilla.jss.crypto.X509Certificate getCertificate(String alias) {
```

Comment:
We can't fix the type problem, so we should just choose one and stick with it. 

--- 

### Rabbit Hole ###

The difference between a `org.mozilla.jss.crypto.X509Certificate` and a [`java.security.cert.X509Certificate`](https://docs.oracle.com/javase/8/docs/api/javax/security/cert/X509Certificate.html) are that the former pre-dates the latter by a bit (~2 years). It also doesn't really contain much and is just an interface. PK11Cert -- where all of the magic happens -- implements both.

The other problem -- and the bigger reason why this wouldn't necessarily work - is that JSS's `X509Certificate` is an interface, but Java's `X509Certificate` is an abstract class. 

So if we make X509Certificate a class and X509Certificate an interface... you get to figure out how to deal with everything. There's a couple of interfaces:

 - `X509Certificate`
 - `InternalCertificate extends X509Certificate`
 - `TokenCertificate extends X509Certificate`

You also have a couple of classes:

 - `PK11Cert extends java.X509Certertificate implements jss.X509Certertificate`
 - `PK11InternalCert extends PK11Cert implements InternalCertificate`

Interfaces can't extend abstract classes. So `X509Certificate` needs to be an abstract class. If you make that a class you have two options:

1. Either `InternalCertificate` and `TokenCertificate` remain interfaces, but drop the inheritance from `jss.X509Certificate`. **This makes them less useful/powerful**, but everything compiles. When you run tests, things seem to work in JSS. I'd bet money though, that we'd have to redo a lot of Dogtag PKI, because there's a bunch of references to `InternalCertificate` and `TokenCertificate` there.
2. Or we make `InternalCertificate` and `TokenCertificate` abstract classes as well. This doesn't work because `PK11InternalCertificate` extends `PK11Cert` and also implements `InternalCertificate`. A class can't extend from two parent classes, so we're stuck.

So 1 is the only viable approach, but would require extending everything to use concrete implementations (`PK11InternalCertificate` and `PK11TokenCertificate`) instead of interfaces. I'm not sure that's a good thing in general.

Coming out of the rabbit hole, we could make JSS's `X509Certificate` have the same signature as Java's, but I'm not sure that gets us anything.

--- 

My argument is that, since this method doesn't exist on Java's `KeyManagers` -- and we specifically want a `JSSKeyManager` instance in `JSSEngine` where this is used -- we should make this return a JSS `X509Certificate`, like I have. This lets us (most safely) cast to `PK11Cert`. There are other Java `X509Certificate` implementations, but only one core JSS `X509Certificate` implementation.

Reply:
FWIW, my response to the second part of your comment is hidden in there:

> There is only one core JSS X509Certificate implementation.

In other words, the cast from `org.mozilla.jss.crypto.X509Certificate` to `PK11Cert` is almost certainly safe. When we get such a `c` that isn't a `PK11Cert`, we want to know what object it is specifically (hence ` + c` on the log statement). I believe that information wouldn't be shown on the cast error, but I could be wrong.

Reply:
Yeah, I'm not suggesting to fix the `X509Certificate` hierarchy now :)
But if we do, I think the `InternalCertificate` and  `TokenCertificate` can be merged into `org.mozilla.jss.crypto.X509Certificate` (or `PK11Cert`) which will then extend `java.security.cert.X509Certificate`.

I'll let you decide on the return type of `getCertificate()`. Even if we pick `org.mozilla.jss.crypto.X509Certificate`, when we eventually fix the hierarchy it shouldn't be much of an issue anymore.

About the type casting, I think something like this will make more sense:
```
logger.debug("JSSKeyManager: - " + c.getSubjectDN());
if (c instanceof PK11Cert) {
    list.add((PK11Cert) c);
} else {
    throw new ClassCastException("Can't cast " + c.getClass().getName() + " to PK11Cert: " + c.getSubjectDN());
}
```
No big deal, I'll leave that up to you. The code looks fine regardless. ACK.

Reply:
ACK, will clean up casting in a separate PR.